src/wayland.rs

```diff
@@ -1,84 +1,201 @@
-use crate::{GraphicsContextImpl, SoftBufferError, error::unwrap};
-use std::{fs::File, os::unix::prelude::{AsRawFd, FileExt}, io::Write};
+use std::{
+    ffi::CStr,
+    fs::File,
+    io::Write,
+    os::unix::prelude::{AsRawFd, FileExt, FromRawFd},
+};
+use wayland_client::{
+    backend::{Backend, ObjectId},
+    globals::{registry_queue_init, GlobalListContents},
+    protocol::{wl_buffer, wl_registry, wl_shm, wl_shm_pool, wl_surface},
+    Connection, Dispatch, EventQueue, Proxy, QueueHandle,
+};
+
+struct State;
 
 pub struct WaylandImpl {
-    _event_queue: EventQueue,
-    surface: WlSurface,
-    shm: Main<WlShm>,
+    event_queue: EventQueue<State>,
+    qh: QueueHandle<State>,
+    surface: wl_surface::WlSurface,
+    shm: wl_shm::WlShm,
     tempfile: File,
-    buffer: Option<WaylandBuffer>
+    buffer: Option<WaylandBuffer>,
 }
 
-struct WaylandBuffer{
+struct WaylandBuffer {
     width: i32,
     height: i32,
-    buffer: Main<WlBuffer>
+    pool: wl_shm_pool::WlShmPool,
+    buffer: wl_buffer::WlBuffer,
 }
 
-impl WaylandImpl {
+impl Drop for WaylandBuffer {
+    fn drop(&mut self) {
+        self.buffer.destroy();
+        self.pool.destroy();
+    }
+}
 
-    pub unsafe fn new<W: HasRawWindowHandle>(window_handle: WaylandWindowHandle, display_handle: WaylandDisplayHandle) -> Result<Self, SoftBufferError<W>> {
-        let display = Display::from_external_display(display_handle.display as *mut wl_display);
-        let mut event_queue = display.create_event_queue();
-        let attached_display = (*display).clone().attach(event_queue.token());
-        let globals = GlobalManager::new(&attached_display);
-        unwrap(event_queue.sync_roundtrip(&mut (), |_, _, _| unreachable!()), "Failed to make round trip to server")?;
-        let shm = unwrap(globals.instantiate_exact::<WlShm>(1), "Failed to instantiate Wayland Shm")?;
-        let tempfile = unwrap(tempfile(), "Failed to create temporary file to store buffer.")?;
-        let surface = Proxy::from_c_ptr(window_handle.surface as _).into();
-        Ok(Self{
-            _event_queue: event_queue,
-            surface, shm, tempfile,
-            buffer: None
+impl WaylandImpl {
+    pub unsafe fn new<W: HasRawWindowHandle>(
+        window_handle: WaylandWindowHandle,
+        display_handle: WaylandDisplayHandle,
+    ) -> Result<Self, SoftBufferError<W>> {
+        let conn = Connection::from_backend(Backend::from_foreign_display(
+            display_handle.display as *mut _,
+        ));
+        let (globals, event_queue) = unwrap(
+            registry_queue_init(&conn),
+            "Failed to make round trip to server",
+        )?;
+        let qh = event_queue.handle();
+        let shm: wl_shm::WlShm = unwrap(
+            globals.bind(&qh, 1..=1, ()),
+            "Failed to instantiate Wayland Shm",
+        )?;
+        let name = CStr::from_bytes_with_nul_unchecked("softbuffer\0".as_bytes());
+        let tempfile_fd = unwrap(
```

Comment:
We should have a fallback or else this only works on Freebsd and Linux. See sctk for the fallback.